src/hotspot/share/opto/compile.cpp

```diff
@@ -3490,7 +3490,13 @@ void Compile::final_graph_reshaping_main_switch(Node* n, Final_Reshape_Counts& f
         } else if (t->isa_oopptr()) {
           new_in2 = ConNode::make(t->make_narrowoop());
         } else if (t->isa_klassptr()) {
-          new_in2 = ConNode::make(t->make_narrowklass());
+          ciKlass* klass = t->is_klassptr()->exact_klass();
+          if (klass->is_interface() || klass->is_abstract()) {
+            // CmpPNode::Ideal should always fold such comparisons
+            assert(false, "Interface or abstract class pointers should not be compressed");
+          } else {
+            new_in2 = ConNode::make(t->make_narrowklass());
```

Comment:
Right, I was hoping for that too and tried to move the assert into `TypeNarrowKlass::make`. We do have all the information there but we hit false positives in rare cases like this when `MyAbstract` does not have any subtypes at compile time (mostly with `-Xcomp`):
```
MyAbstract obj = ...;
obj.getClass();
```
C2 will add a dependency that will invalidate the code once a subclass is loaded and then optimizes the narrow class load from `obj` to be of constant narrow class type `MyAbstract`. The assert will trigger but we will never emit a compressed class pointer because the narrow class load + decode is folded to a non-narrow constant.

We could move the assert to a later stage though. I'll give that a try.

Reply:
Do we actually generate an nmethod for the above example?  It seems like it could never execute the getClass() because the line above setting `obj` would have to throw an exception if there can be no concrete instances.

Reply:
Right, this was an oversimplified example. I used this code:
```
Class test(MyAbstract obj, boolean b) {
  if (b) {
    return obj.getClass();
  }
  return null;
}
```

We pass `null` for `obj` and `false` for `b`. Usually, the branch is then only compiled with Xcomp.

Reply:
I think there is still hope for moving the assert into `TypeNarrowKlass::make` in a future RFE.  In the example above, if we are generating code for obj.getClass() based on the assumption that the type is a leaf, we could also notice that the type is abstract and deduce that obj must be null.

Reply:
Right, we could do that.